### PR TITLE
Feat: XMR subaddresses per account

### LINKED ISSUE #4500: Reuse of Monero receiving address
In bisq, each Monero account supports a single Monero receiving address. Even though reusing a Monero address is not as dangerous as reusing a Bitcoin address (because Monero address is never stored in the public blockchain), this practice can still compromise user's privacy if there is off-chain linking.

Instead of asking for a single receiving Monero address, the Monero account in bisq could ask for the public spend key and the private (secret) view key of the Monero wallet. By having these two keys, bisq would be able to generate a new subaddress for each new transaction, avoiding address reuse and increasing user privacy. Take a look at this javascript solution in https://github.com/knaccc/subaddress-js.

### PR DESCRIPTION
**Pull request by ghost**
Fixes: #4500

Account creation:
xmr_subaddress_creation_danda

The subaddress field is auto-generated when all other fields have valid data entered, otherwise it is blank and the account cannot be saved.  Example test data, from https://github.com/dan-da/subaddress-derive-xmr :

    wallet address: 49zf2PF7nLSHpRwWcPG8ePHxYnR6eFmYuKG8Akpq5vFALTzZzMdv3kC36fCSP3UfFdMrY51QAs5NGiGuwXK6YMa3Nk7549x
    view-key-private: 25d014a444fb7a1e6836c680d3ec1b6eed628a29c3c85e0379fb89f53c4c610a
    account 0 subaddress 1 87i7kA61fNvMboXiYWHVygPAggKJPETFqLXXcdH4mQTrECvrTxZMtt6e6owj1k8jUVjNR11eBuBMWHFBtxAwEVcm9dcSUxr |
    account 0 subaddress 2 8A9XmWsATrhfedtNhTMNKELwfCwMVAk2iVTdUJdFRb2AC4tV4VeBjsCLYR9cSQTwnvLo4MAuQFMLP6Si4xp6t6BS788db3t |

Two simultaneous trades.  Notice the subaddresses match the example above.

XMR seller(s):
image

XMR buyer:
image

---

&nbsp;

initial account holder notice (text rewritten by chimp1984)

image

&nbsp;

info at account creation when subaddress checkbox selected (text rewritten by chimp1984)

image

&nbsp;

account detail showing list of subaddresses used and their trades

image

### CODE REVIEW

**Comment by ghost** [comment]
@chimp1984 
The XMR payment account payload is processed [at the end of trade initialization](https://github.com/bisq-network/bisq/pull/6236/files#diff-d5acf1c920949c3b8ddb31fcc71641b243f049eb90d59e9865d326052dea6025) after payment account payloads have been delivered immutably to the contract.  Once that is done, we're free to bump the subaddress index for use in the next trade.  However, there's a window of time before that.  It would be better to store the buyer's payment account payload earlier so that window is minimized, reducing the risk of simultaneous trades claiming the same subaddress.

File: core/src/main/java/bisq/core/payment/XmrAccountHelper.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    public void setSubAddressIndex(String newValue) {
+        account.getOrCreateExtraData().put(KEY_SUB_ADDRESS_INDEX, newValue);
+    }
+
+    public void recalculateSubAddress() {
+        try {
+            long accountIndex = Long.parseLong(getAccountIndex());
+            long subAddressIndex = Long.parseLong(getSubAddressIndex());
+            if (subAddressIndex < 0 || accountIndex < 0 || (subAddressIndex + accountIndex < 1)) {
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/payment/XmrAccountHelper.java]: Is it not permitted that both are 0?

File: core/src/main/java/bisq/core/payment/XmrAccountHelper.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                throw new NumberFormatException();  // prevent invalid subaddress/account index
+            }
+            account.getOrCreateExtraData().put(KEY_SUB_ADDRESS_INDEX, Long.toString(subAddressIndex));
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/payment/XmrAccountHelper.java]: Why is it set here we are reading it a few lines above?

File: core/src/main/java/bisq/core/payment/XmrAccountHelper.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            if (getMainAddress().length() > 0 && getPrivateViewKey().length() > 0) {
+                WalletAddress walletAddress = new WalletAddress(getMainAddress());
+                String subaddress = walletAddress.getSubaddressBase58(getPrivateViewKey(), accountIndex, subAddressIndex);
+                account.setAddress(subaddress);
+                return;
+            }
+        } catch (Exception e) {
+            log.warn(e.toString());
+        }
+        account.setAddress("");
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/payment/XmrAccountHelper.java]: Maybe better move it into the catch to make it more clear.

File: core/src/main/java/bisq/core/payment/XmrAccountHelper.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    public static void createNewSubAddress(CryptoCurrencyAccount cryptoCurrencyAccount) {
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/payment/XmrAccountHelper.java]: I find it better if static methods are at the top of the class...

File: core/src/main/resources/i18n/displayStrings.properties [hunk 5]
```diff
@@ -1648,10 +1648,15 @@ but you need to enable it in Settings.\n\n\
 See the wiki for more information about the auto-confirm feature: [HYPERLINK:https://bisq.wiki/Trading_Monero#Auto-confirming_trades].\n\n\
 Please be aware that when changing the Bisq data directory you must change the Monero wallet and address to avoid \
   vulnerabilities related to transaction key re-use. More details can be found here: [HYPERLINK:https://www.getmonero.org/2018/09/25/a-post-mortum-of-the-burning-bug.html].
-
-account.altcoin.popup.xmr.dataDirWarning=Please be aware that when changing the Bisq data directory you must change the Monero wallet and address to avoid \
-  vulnerabilities related to transaction key re-use. More details can be found here: [HYPERLINK:https://www.getmonero.org/2018/09/25/a-post-mortum-of-the-burning-bug.html].
-
+account.altcoin.popup.xmr.dataDirWarningHeadline=Notice for XMR users
```
Comment (chimp1984) [inline core/src/main/resources/i18n/displayStrings.properties]: I am not convinced that this it more clear. It lacks information why its a security issue and mixes in another issue with changing data directories. 
Maybe @m52go could have a look into it.

File: core/src/main/resources/i18n/displayStrings.properties [hunk 6]
```diff
@@ -1648,10 +1648,15 @@ but you need to enable it in Settings.\n\n\
...
+account.altcoin.popup.xmr.dataDirWarning=If changing data directory you must also change XMR wallet and address for security reasons \
+  [HYPERLINK:https://www.getmonero.org/2018/09/25/a-post-mortum-of-the-burning-bug.html]. \
+  Read about why it's a good idea to switch to a new data directory every so often \
+  [HYPERLINK:https://bisq.wiki/Switching_to_a_new_data_directory].
+account.altcoin.popup.xmr.subAddressHeadline=Monero Subaddresses
+account.altcoin.popup.xmr.subAddressInfo=Information about monero subaddresses can be found here: [HYPERLINK:https://monerodocs.org/public-address/subaddress]\n\n\
+  Your main XMR wallet address and View Key [HYPERLINK:https://www.getmonero.org/resources/user-guides/view_only.html] are needed for Bisq to derive a new subaddress per trade. \
+  It is highly recommended you verify that the subaddress shown in the Bisq account matches your monero wallet before trading.
```
Comment (chimp1984) [inline core/src/main/resources/i18n/displayStrings.properties]: Can the user see the same subaddress in their XMR wallet? I assume the incremented index is not editable in the account UI.

File: desktop/src/main/java/bisq/desktop/components/paymentmethods/XmrForm.java [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+        regexValidator.setErrorMessage(Res.get("portfolio.pending.step2_buyer.confirmStart.proof.invalidInput"));
+        privateViewKeyInputTextField.setValidator(regexValidator);
+        privateViewKeyInputTextField.textProperty().addListener((ov, oldValue, newValue) -> {
+            updateFromInputs();
+        });
+
+        HBox hBox = new HBox();
+        hBox.setSpacing(10);
+
+        //accountIndex = FormBuilder.addInputTextField(gridPane, ++gridRow, Res.get("payment.altcoin.xmrAccountIndex"));
```
Comment (chimp1984) [inline desktop/src/main/java/bisq/desktop/components/paymentmethods/XmrForm.java]: Can we remove the commented out code lines?

File: desktop/src/main/java/bisq/desktop/components/paymentmethods/XmrForm.java [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+                xmrAccountHelper.setSubAddressIndex(subAddressIndex.getText());
+                xmrAccountHelper.recalculateSubAddress();
+                subAddressTextField.setText(xmrAccountHelper.getSubAddress());
+            } else {
+                xmrAccountHelper.setSubAddress("");
+                subAddressTextField.setText("");
+            }
+        } else {
+            // legacy XMR (no subaddress)
+            xmrAccountHelper.setSubAddress(addressInputTextField.getText());
```
Comment (chimp1984) [inline desktop/src/main/java/bisq/desktop/components/paymentmethods/XmrForm.java]: Do we use the xmrAccountHelper.subAddress field for the main address in that case? I think thats confusing.

File: core/src/main/java/bisq/core/payment/PaymentAccount.java [hunk 9]
```diff
@@ -130,6 +137,14 @@ public static PaymentAccount fromProto(protobuf.PaymentAccount proto, CoreProtoR
             if (proto.hasSelectedTradeCurrency())
                 account.setSelectedTradeCurrency(TradeCurrency.fromProto(proto.getSelectedTradeCurrency()));
 
+            if (CollectionUtils.isEmpty(proto.getExtraDataMap())) {
+                account.setExtraData(null);
+            } else {
+                Map<String, String> clonedExtra = new HashMap<>();
+                Map<String, String> extraData = proto.getExtraDataMap();
+                extraData.forEach(clonedExtra::putIfAbsent);
+                account.setExtraData(clonedExtra);
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/payment/PaymentAccount.java]: Could be simplified with `account.setExtraData(new HashMap<>(proto.getExtraDataMap()));`

File: core/src/main/java/bisq/core/payment/XmrAccountHelper.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        XmrAccountHelper helper = new XmrAccountHelper(cryptoCurrencyAccount);
+        Optional.ofNullable(helper.getAccount().extraData).ifPresent(e -> {
+            if (helper.getPrivateViewKey() != null && helper.getPrivateViewKey().length() > 0) {
```
Comment (chimp1984) [inline core/src/main/java/bisq/core/payment/XmrAccountHelper.java]: helper.getPrivateViewKey() is never null

**Review by chimp1984 — CHANGES_REQUESTED** [review]

**Comment by chimp1984** [comment]
Sent a suggested patch with further changes. I think we should increment the index at the start of the trade as that causes less mutation issues.

**Comment by ghost** [comment]
Closing: chimp1984 will rewrite.

**Comment by ripcurlx** [comment]
@chimp1984 As I'm planing to prepare testing for v1.9.2 soon. Do you have a timeline for this change, so we can squeeze it into the next update? Thanks!

**Comment by chimp1984** [comment]
@jmacxx has to decide if it can make it into that release. I would be good as then we only need to tell the user once about the issue and subaddresses help greatly to reduce the risk.

**Review by chimp1984 — CHANGES_REQUESTED** [review]
Looks all good but I send @jmacxx a patch for some suggested changes. Also we should promote it with a NEW badge and remove the BSQ swap badge.

**Review by chimp1984 — DISMISSED** [review]
utACK

**Review by chimp1984 — DISMISSED** [review]
utACK

**Comment by ghost** [comment]
This popup shown at `Account` is the same as the one shown when Bisq is started.  Feels a bit repetitive to show it twice. 

We do need to show something at Account, as we've added the News badge.

:thinking:   :grey_question: 


image

**Comment by chimp1984** [comment]
> This popup shown at `Account` is the same as the one shown when Bisq is started. Feels a bit repetitive to show it twice.
> 
> We do need to show something at Account, as we've added the News badge.
> 
> 🤔 ❔
> 
> image

Yes, I guess we can remove the start-up popup.

**Review by chimp1984 — DISMISSED** [review]
utACK

**Review by ripcurlx — DISMISSED** [review]
utACK - based on https://github.com/bisq-network/bisq/pull/6236#pullrequestreview-1004384704

**Comment by ripcurlx** [comment]
@jmacxx It looks like the cherry-pick commits are not signed properly. Could you force push those changes?

**Comment by ghost** [comment]
Removed the penalties popup (and no longer used Bisq Swaps popup text), per https://github.com/bisq-network/bisq/discussions/6244#discussioncomment-2933890

**Review by ripcurlx — APPROVED** [review]
utACK - based on https://github.com/bisq-network/bisq/pull/6236#pullrequestreview-1004384704

**Pull request merged**